### PR TITLE
Provide access to unwrapped model in Lite

### LINKED ISSUE #12596: Provide access to unwrapped module in Lite
## 🚀 Feature

LightningLite has a method `setup` which wraps the model:

```py
model, optimizer = self.setup(model, optimizer)
```

Here model is an instance of `_LiteModule` and to access the attributes of the original model, you would do
`model.module.method_a()` instead of `model.method_a()`. 

However, depending on the Strategy, the model may be wrapped by other modules like `DistributedDataParallel` for example.

### Motivation

Normally one only calls the model through forward. However, in some cases it is useful to expose extra method on the model that don't do distributed communication. 

### Pitch

Make `_LiteModule.module` return the original unwrapped model, as it was passed into the setup() method.

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @carmocca @justusschock @awaelchli

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Fixes #12596

This PR enables attribute pass-through on the _LiteModule wrapper. When an attribute is not found on the wrapper, it tries to look it up on the original model that was passed into the setup method.

The _LiteModule wrapper now takes an optional argument `original_module`, which is the one passed in to the `Lite.setup` method. If we didn't do this, we wouldn't be able to access it, because the `forward_module` is potentially already wrapped by something like `DistributedDataParallel`.

## Breaking change

This is a breaking change if you previously relied on `LiteModule.state_dict()` to save or load checkpoints. The old checkpoint won't load after this PR because of missing keys (_original_module, _forward_module).

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

cc @borda @tchaton @rohitgr7 @carmocca @justusschock @awaelchli

### CODE REVIEW

File: pytorch_lightning/lite/wrappers.py [hunk 1]
```diff
@@ -65,23 +65,27 @@ def step(self, closure: Optional[Callable] = None) -> Any:
...
         super().__init__()
-        self._module = module
+        self._forward_module = forward_module
+        self._original_module = original_module
         self._precision_plugin = precision_plugin
 
     @property
     def module(self) -> nn.Module:
-        return self._module
+        return self._original_module or self._forward_module
```
Comment (justusschock) [inline pytorch_lightning/lite/wrappers.py:88]: Should we really have a separate reference here? Can't we just call the unwrap function, we used to have?

Comment (awaelchli) [reply pytorch_lightning/lite/wrappers.py:90]: The "unwrap" function is implemented for LightningModule and has a stopping condition with an isinstance check against LightningModule. This can't be so easily converted to an equivalent unwrapping method for pure nn.Modules.
Afaik, this is the safest way to guarantee we actually obtain the original untouched model that was passed in to the Lite.setup method.

**Review by Borda — APPROVED** [review]

File: pytorch_lightning/lite/wrappers.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/lite/wrappers.py:90]: Off topic to the PR: I find the unwrapping logic we have in Lightning hard to follow. I would like that we clean it up someday

This solution by Adrian is clean and explicit IMO

File: pytorch_lightning/lite/wrappers.py [hunk 2]
```diff
@@ -102,12 +108,21 @@ def _convert_float_tensor(t: Tensor) -> Tensor:
...
         to_type = torch.get_default_dtype()
         output = apply_to_collection(output, function=_convert_float_tensor, dtype=Tensor)
         return output
 
+    def __getattr__(self, item: Any) -> Any:
+        try:
+            # __getattr__ gets called as a last resort if the attribute does not exist
+            # call nn.Module's implementation first
+            return super().__getattr__(item)
+        except AttributeError:
```
Comment (carmocca) [inline pytorch_lightning/lite/wrappers.py:122]: Why the try-except instead of `hasattr`?

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,7 +9,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
--
+- Added attribute lookup pass-through for the wrapped model returned by `LightningLite.setup()` ([#12597](https://github.com/PyTorchLightning/pytorch-lightning/pull/12597))
```
Comment (carmocca) [inline CHANGELOG.md:12]: If this is set to the 1.6.x milestone, shouldn't this be in "Fixed"?

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/lite/wrappers.py (hunk 2 above)
Comment (justusschock) [reply pytorch_lightning/lite/wrappers.py:122]: Afaik `hasattr` is implemented by calling `getattr` and catching `AttributeError`. So if you would call `hasattr` here, `getattr` would call this very method again and you'd end up with recursion errors since you can only call this on an instance (`self`) and not the superclass

File: pytorch_lightning/lite/wrappers.py (hunk 1 above)
Comment (justusschock) [reply pytorch_lightning/lite/wrappers.py:90]: I agree, The only motivation for asking was consistency so that when we clean it in one place, it would still all behave the same :)

File: pytorch_lightning/lite/wrappers.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/lite/wrappers.py:122]: Yes exactly, that is the reason. 
However, there is still something weird going on with one of the failing deepspeed test and haven't figured it out yet. 

```
tests/lite/test_lite.py:459: 
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ 
pytorch_lightning/lite/lite.py:405: in _run_impl
    return self._strategy.launcher.launch(run_method, *args, **kwargs)
pytorch_lightning/strategies/launchers/subprocess_script.py:93: in launch
    return function(*args, **kwargs)
pytorch_lightning/lite/lite.py:412: in _run_with_strategy_setup
    return run_method(*args, **kwargs)
tests/lite/test_lite.py:402: in run
    model, optimizer = self.setup(model, optimizer)
pytorch_lightning/lite/lite.py:174: in setup
    model = _LiteModule(model, self._precision_plugin, original_module=original_model)
/usr/local/lib/python3.7/dist-packages/deepspeed/runtime/zero/partition_parameters.py:249: in wrapper
    if not hasattr(module, "_ds_child_entered"):
pytorch_lightning/lite/wrappers.py:124: in __getattr__
    return getattr(self._original_module, item)
pytorch_lightning/lite/wrappers.py:124: in __getattr__
    return getattr(self._original_module, item)
pytorch_lightning/lite/wrappers.py:124: in __getattr__
    return getattr(self._original_module, item)
E   RecursionError: maximum recursion depth exceeded while calling a Python object
!!! Recursion detected (same locals & position)
```

My suspicion is that deepspeed does some patching which I haven't found yet.

Comment (awaelchli) [reply pytorch_lightning/lite/wrappers.py:122]: Ok, finally figured this out. Deepspeed patches the init globally for all nn Modules. That wrapper called hasattr which then called our getattr and at this point `_original_module ` wasn't defined yet so it would recurse infinitely.

**Review by akihironitta — COMMENTED** [review]
lgtm 🚀

**Review by otaj — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by rohitgr7 — APPROVED** [review]

**Comment by awaelchli** [comment]
Gatekeeper merge
